chain/epoch_manager/src/lib.rs

```diff
@@ -1082,6 +1082,24 @@ impl EpochManager {
         let stake_divisor = self.config.minimum_stake_divisor as Balance;
         Ok(self.get_epoch_info(&next_epoch_id)?.seat_price() / stake_divisor)
     }
+
+    // Note: this function should only be used in 18 -> 19 migration and should be removed in the
+    // next release
+    /// `block_header_info` must be the header info of the last block of an epoch.
+    pub fn migrate_18_to_19(
```

Comment:
@matklad btw I wonder whether there is any better way to do this. Essentially I need a non-public function in some migration code that only needs to be run once. Current plan is to remove this code later

Reply:
I don't think there's any specific great solution here. Though, I'd probably just make `finalize_epoch` public (with a comment), such that I could write migration code in the migration file. I can also imagine some kind of `#[cfg(feature = "migration_api")]`, but that'd be overkill -- we have enough features already :) 

Although there's one thing I don't quite understand. From the reading of it, we have all the migrations, and in theory can migrate 1 database all the way to 19. Does removing this code in the future also mean that we forgo the ability to forward-migrate version 1 database?

Reply:
> Does removing this code in the future also mean that we forgo the ability to forward-migrate version 1 database?

I don't think that is necessarily an issue because there is no need for that. You only need forward compatibility if for example, you expect that you can run the code corresponds to database version 1 and it still works, but that is simply not the case. The protocol version and changed in between and if you run a node with some commit that corresponds to database version 1, it just won't be able to join the network.

Reply:
Yeah, I don't have an opinion if this is an issues or not :)

I was just surprised to see that we seemingly keep ability to migrate from version 1, while discussing removing the ability to migrate from version 18 some time in the future.  

If we dont't need to keep comapt with DB version 1 forever, than the problem indeed is much easier, and temporary function seems to be a perfect solution.